web manifest: implement orientation member parsing

marcoscaceres opened the pull request:
#### f3713b3399174f5ce5a32b1e0b8b99a5f9956c7d
<pre>
web manifest: implement orientation member parsing
<a href="https://bugs.webkit.org/show_bug.cgi?id=247275">https://bugs.webkit.org/show_bug.cgi?id=247275</a>
rdar://101770484

Reviewed by Chris Dumez.

* Source/WebCore/Modules/applicationmanifest/ApplicationManifest.h:
* Source/WebCore/Modules/applicationmanifest/ApplicationManifestParser.cpp:
(WebCore::ApplicationManifestParser::parseManifest):
(WebCore::ApplicationManifestParser::parseOrientation):
* Source/WebCore/Modules/applicationmanifest/ApplicationManifestParser.h:
* Source/WebCore/page/ScreenOrientationLockType.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/API/Cocoa/_WKApplicationManifest.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKApplicationManifest.mm:
(-[_WKApplicationManifest initWithCoder:]):
(-[_WKApplicationManifest encodeWithCoder:]):
(-[_WKApplicationManifest orientation]):
* Tools/TestWebKitAPI/Tests/WebCore/ApplicationManifestParser.cpp:
(WebCore::operator&lt;&lt;):
(ApplicationManifestParserTest::testOrientation):
(TEST_F):

Canonical link: <a href="https://commits.webkit.org/262241@main">https://commits.webkit.org/262241@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/20c856d532e96636254dc0468c61d60a035d1908

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/899 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/923 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/958 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/1276 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/795 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/887 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/973 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/1008 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/999 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/907 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/844 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/842 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/1211 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/897 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/837 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/835 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/814 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/863 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/1863 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/855 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/809 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/802 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/848 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/239 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/869 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/page/ScreenOrientationLockType.h:
```diff
@@ -37,7 +37,8 @@ enum class ScreenOrientationLockType : uint8_t {
     LandscapePrimary,
-    LandscapeSecondary
+    LandscapeSecondary,
+    Undefined
```
I don’t think we should be adding an enum value. Have your parser return a std::optional if you don’t want it to return ANY in case of parsing failure.

marcoscaceres replied:
Ok, yeah. I saw there was other cases that were using `Undefined` so copied that. std::optional would be better.

cdumez commented on Source/WebCore/page/ScreenOrientationLockType.h:
```diff
@@ -48,13 +49,14 @@ template<> struct EnumTraits<WebCore::ScreenOrientationLockType> {
         WebCore::ScreenOrientationLockType::Any,
-        WebCore::ScreenOrientationLockType::Natural,
         WebCore::ScreenOrientationLockType::Landscape,
+        WebCore::ScreenOrientationLockType::LandscapePrimary,
```
Why are these enum traits still here if you are also adding the enum to the serialization.in. As I mentioned in a previous review, it is one of the other.

also the changes to this enum traits don’t seem right. As far as I know, the ordering needs to match the enum’s ordering.

marcoscaceres replied:
Sorry about that! Not sure what happened there. I thought I deleted these :/

cdumez commented on Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
```diff
@@ -770,6 +770,18 @@ struct WebCore::ApplePayCouponCodeUpdate : WebCore::ApplePayDetailsUpdateBase {
+header: <WebCore/ScreenOrientationLockType.h>
+enum class WebCore::ScreenOrientationLockType : uint8_t {
+    Any
+    Landscape
```
As mentioned in an earlier review, this should match the ordering of the enum as far as I know. Otherwise I wouldn't be surprised if encoding / decoding of the enum was completely wrong.

marcoscaceres replied:
oops, sorry. I'd assumed the enum was also alphabetically ordered.

cdumez replied:
The whole point of this is to serialize / deserialize for IPC. It likely gets send as an unsigned integer and thus ordering is important if you want the same enum value on the other side of the pipe.

marcoscaceres replied:
yeah, makes sense. I'll be sure to check these more carefully next time.

cdumez commented on Tools/TestWebKitAPI/Tests/WebCore/ApplicationManifestParser.cpp:
```diff
@@ -106,6 +129,20 @@ class ApplicationManifestParserTest : public testing::Test {
         EXPECT_EQ(expectedValue, value);
     }
 
+    void testOptionalOrientation(const String& rawJSON)
```
I don't think this function is needed.

cdumez commented on Tools/TestWebKitAPI/Tests/WebCore/ApplicationManifestParser.cpp:
```diff
@@ -305,6 +342,27 @@ TEST_F(ApplicationManifestParserTest, Display)
 
+TEST_F(ApplicationManifestParserTest, Orientation)
+{
+    testOptionalOrientation(""_s);
```
These functions could just call:
`testOrientation("123"_s, std::nullopt);`

cdumez commented on Source/WebCore/Modules/applicationmanifest/ApplicationManifestParser.h:
```diff
@@ -46,6 +47,7 @@ class ApplicationManifestParser {
 
     URL parseStartURL(const JSON::Object&, const URL&);
     ApplicationManifest::Display parseDisplay(const JSON::Object&);
+    std::optional<WebCore::ScreenOrientationLockType> parseOrientation(const JSON::Object&);
```
We're in WebCore so no `WebCore::`

cdumez replied:
Can this function be marked as `const` or is there something that present it?

marcoscaceres replied:
Yes, I think it could. 

I think the other parser method s too could be `const`. I'll do the others as a followup.

cdumez commented on Source/WebCore/Modules/applicationmanifest/ApplicationManifestParser.cpp:
```diff
@@ -162,6 +164,38 @@ ApplicationManifest::Display ApplicationManifestParser::parseDisplay(const JSON:
     return ApplicationManifest::Display::Browser;
 }
 
+std::optional<WebCore::ScreenOrientationLockType> ApplicationManifestParser::parseOrientation(const JSON::Object& manifest)
```
No `WebCore::`

cdumez reviewed (approved):
LGTM with minor fixes.

marcoscaceres committed:
web manifest: implement orientation member parsing
https://bugs.webkit.org/show_bug.cgi?id=247275
rdar://101770484

Reviewed by Chris Dumez.

* Source/WebCore/Modules/applicationmanifest/ApplicationManifest.h:
* Source/WebCore/Modules/applicationmanifest/ApplicationManifestParser.cpp:
(WebCore::ApplicationManifestParser::parseManifest):
(WebCore::ApplicationManifestParser::parseOrientation):
* Source/WebCore/Modules/applicationmanifest/ApplicationManifestParser.h:
* Source/WebCore/page/ScreenOrientationLockType.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/API/Cocoa/_WKApplicationManifest.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKApplicationManifest.mm:
(-[_WKApplicationManifest initWithCoder:]):
(-[_WKApplicationManifest encodeWithCoder:]):
(-[_WKApplicationManifest orientation]):
* Tools/TestWebKitAPI/Tests/WebCore/ApplicationManifestParser.cpp:
(WebCore::operator<<):
(ApplicationManifestParserTest::testOrientation):
(TEST_F):

Canonical link: https://commits.webkit.org/262241@main
